### PR TITLE
Encoder module!

### PR DESCRIPTION
**Pull request by pullenrc**
Here is my stab at a module for encoders.  It is tested a and working on an Atreus62 with a single encoder.  I really like the modular design of this new branch, this was pretty easy to do once I got my head wrapped around it.

### CODE REVIEW

File: boards/atreus62/README.md [hunk 1]
```diff
@@ -0,0 +1,18 @@
...
+
+Retailers (USA)  
+[Atreus62](https://shop.profetkeyboards.com/product/atreus62-keyboard)  
+
+Extentions enabled by default  
+- [Layers](https://github.com/KMKfw/kmk_firmware/tree/master/docs/layers.md) Need more keys than switches? Use layers.
+- [RGB](https://github.com/KMKfw/kmk_firmware/tree/master/docs/rgb.md) Light it up
+- [Encoder](https://github.com/KMKfw/kmk_firmware/tree/master/docs/encoder.md) Twist control for all the things
+
+Common Extentions
```
Comment (kdb424) [inline boards/atreus62/README.md:17]: If this section is not needed, it can be removed.

File: docs/encoder.md [hunk 2]
```diff
@@ -0,0 +1,48 @@
+# Encoder
+Add twist control to your keyboard! Volume, zoom, anything you want.
+
+## Enabling the extension
+The constructor takes a minimun of 3 arguments: a list of pad_a pins, a list of pad_b pins, 
+and an encoder_map.  The encoder_map is modeled after the keymap and works the
+same way. It shoul have as many layers as your keymap, and use filler keys for
```
Comment (kdb424) [inline docs/encoder.md:7]: `shoul` should be `should`. Typo

File: docs/encoder.md [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+Add twist control to your keyboard! Volume, zoom, anything you want.
+
+## Enabling the extension
+The constructor takes a minimun of 3 arguments: a list of pad_a pins, a list of pad_b pins, 
+and an encoder_map.  The encoder_map is modeled after the keymap and works the
+same way. It shoul have as many layers as your keymap, and use filler keys for 
+layers that you don't require any action.  The encoder supports a velocity mode
+if you desire to make something for video or sound editing. The direction of 
+increment/decrement can be changed to make sense for the direction the knob is 
+turning by setting the is_inverted flag.
```
Comment (kdb424) [inline docs/encoder.md:11]: Maybe add `is_inverted` into a code block to make that clear it means something in code.

File: docs/encoder.md [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+
+## Configuration
+Create your special keys:
+```
```
Comment (kdb424) [inline docs/encoder.md:15]: Any python code blocks should start with 
\`\`\`python

in order to get syntax highlighting.

**Review by kdb424 — CHANGES_REQUESTED** [review]
Once those cleanups are made, and I didn't mark repeats, I'll re-review. Good over all though, and debug should be finished or removed before final review as well. Overall, looks great.

File: kmk/modules/encoder.py [hunk 5]
```diff
@@ -0,0 +1,221 @@
...
+            )
+        return keyboard
+
+    def get_reports(self, keyboard):
+        for idx in range(self.encoder_count):
+            if self.debug_enabled: # not working as inttended, do not use for now
+                print(self.encoders[idx].__repr__(idx))
+            encoder_key = self.encoders[idx].report()
+            if encoder_key is not None:
+                k = self.send_encoder_keys(keyboard, encoder_key, idx)
```
Comment (kdb424) [inline kmk/modules/encoder.py:220]: To make this cleaner, why not just return, as opposed to storing?

File: kmk/modules/encoder.py [hunk 6]
```diff
@@ -0,0 +1,221 @@
...
+
+    def __init__(self, pad_a, pad_b, encoder_map):
+        self.pad_a = pad_a
+        self.pad_b = pad_b
+        self.encoder_count = len(self.pad_a)
+        self.encoder_map = encoder_map
+        self.make_encoders()
+
+    def on_runtime_enable(self, keyboard):
+        raise NotImplementedError
```
Comment (kdb424) [inline kmk/modules/encoder.py:159]: These should be blank `return` as opposed to raising an error if they are intentionally not used. This is to catch you if you flat out forget to include them.

File: kmk/modules/encoder.py [hunk 7]
```diff
@@ -0,0 +1,221 @@
...
+
+    def on_runtime_disable(self, keyboard):
+        raise NotImplementedError
```
Comment (kdb424) [inline kmk/modules/encoder.py:162]: Here as well

File: kmk/modules/encoder.py [hunk 8]
```diff
@@ -0,0 +1,221 @@
...
+
+    def during_bootup(self, keyboard):
+        raise NotImplementedError
```
Comment (kdb424) [inline kmk/modules/encoder.py:165]: And here

File: kmk/modules/encoder.py [hunk 9]
```diff
@@ -0,0 +1,221 @@
...
+
+        modified_keyboard = self.get_reports(keyboard)
+
+        return modified_keyboard
+
+    def after_matrix_scan(self, keyboard):
+        '''
+        Return value will be replace matrix update if supplied
+        '''
+        raise NotImplementedError
```
Comment (kdb424) [inline kmk/modules/encoder.py:180]: And the rest. I won't mark them all. You get it.

File: kmk/modules/encoder.py [hunk 10]
```diff
@@ -0,0 +1,221 @@
...
+    # checks encoder pins, reports encoder data
+    def report(self):
+        new_encoder_state = (
+            self.encoder_pad_lookup[int(self.pad_a.value)],
+            self.encoder_pad_lookup[int(self.pad_b.value)],
+        )
+
+        if self.encoder_state == (self.eps.ON, self.eps.ON):  # Resting position
+            if new_encoder_state == (self.eps.ON, self.eps.OFF):  # Turned right 1
+                self.encoder_direction = 'R'
```
Comment (kdb424) [inline kmk/modules/encoder.py:81]: Could we use a bool for this, and document which is which? That should take less resources, even if only by a bit.

Comment (kdb424) [reply kmk/modules/encoder.py:81]: Could also do something like this with a `const`. That will be easier to read, and low memory. `from micropython import const`

**Comment by kdb424** [comment]
Look like a `make fix-formatting fix-isort` is in order as well.

File: docs/encoder.md (hunk 2 above)
Comment (kdb424) [reply docs/encoder.md]: `minimun` to `minimum` as well.

File: docs/encoder.md (hunk 3 above)
Comment (pullenrc) [reply docs/encoder.md:11]: Fixed, I think.

File: docs/encoder.md (hunk 4 above)
Comment (pullenrc) [reply docs/encoder.md]: Added to all blocks of python codee

**Comment by pullenrc** [comment]
Working to get this all fixed this week, thanks for the feedback.

**Review by kdb424 — APPROVED** [review]
Approving this. I'll tag @klardotsh for final code review, but looks great to me.

**Comment by kdb424** [comment]
@pullenrc https://github.com/KMKfw/kmk_firmware/pull/211/checks?check_run_id=3059690312 Can you check the liner errors here, and resolve those. I'll get a merge in as soon as that passes checks.

**Pull request merged**